torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,278 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
       TransformRFactor::runReplay2(this, axes)};
 }
 
+namespace {
+
+/** \brief Container class DisjointSet models equivalence relationships
+ *  \details each instance of this class keeps a set of equivalent classes
+ *           DisjointSet::join(a,b) makes the full class of a and b equivalent
+ *           DisjointSet::areEqual(a,b) checks if a and b belong same class
+ *
+ */
+template <typename T>
+class DisjointSet {
+ public:
+  DisjointSet() = default;
+
+  /** \brief Joins the equivalent class that a and b belong to
+   *         areEqual(a',b') will be true for each a'=a and b'=b
+   */
+  void join(T a, T b) {
+    int i0 = fixedPoint(a), i1 = fixedPoint(b);
+    int new_parent, new_child;
+
+    // either order here is correct but joining larger class to smaller class
+    // tend to be faster
+    std::tie(new_parent, new_child) = (weights[i0] < weights[i1])
+        ? std::make_pair(i0, i1)
+        : std::make_pair(i1, i0);
+    weights[new_parent] += weights[new_child];
+    setMap[new_child] = new_parent;
+  }
+
+  /** \brief Checks if a and b belong to the same equivalent class */
+  bool areEqual(T a, T b) {
+    if (!entryMap.count(a) || !entryMap.count(b))
+      return false;
+    return (fixedPoint(a) == fixedPoint(b));
+  }
+
+ private:
+  // Internal representation of the equivalence class as integers
+  // setMap implements the "parent" relationship
+  // weights is used for preliminary perf optimization
+  std::vector<int> setMap, weights;
+
+  // Map the input of type T to its equivalence class
+  std::unordered_map<T, int> entryMap;
+
+  // Utility for generating new class
+  int count = 0;
+
+  // internal fixed point implementation:
+  //  returns the equivalent class that e belongs to
+  //  does adhoc optimization to speed up future access
+  int fixedPoint(int e) {
+    TORCH_INTERNAL_ASSERT(setMap.size() > e);
+    while (setMap[e] != e)
+      e = setMap[e] = setMap[setMap[e]];
+    return e;
+  }
+
+  // utility to check the class i belongs to:
+  //  returns the equivalent class that e belongs to
+  //  will create a new class if no match seen
+  int fixedPoint(T i) {
+    if (!entryMap.count(i))
+      createPoint(i);
+    return fixedPoint(entryMap[i]);
+  }
+
+  // utility to create a new equiv class for i
+  void createPoint(T i) {
+    entryMap[i] = count;
+    setMap.push_back(count++);
+    weights.push_back(1);
+  }
+}; // class DisjoinSet
```

Comment:
we don't use class ending comments (the only ending scope comments are used for namespaces)

Reply:
Updated. Thanks!